### DISCUSSION TITLE
Working on Simple Validate 2.0

### DISCUSSION THREAD

**Discussion by colinaut** [colinaut/alpinejs-plugin-simple-validate#24] [Announcements]
Hi all,

I'm working on a version 2.0 of this plugin — integrating some enhancements and cleaning up the code to make it more smaller and more maintainable. 

## Enhancements

1. Revamp how checkbox and radio button groups work so that x-validate.group is not required.
   * I always hated how groups worked and feel there is a better simpler way
   * Current thought is to use `data-group-min` attribute on the common parent element, which will also allow for `data-group-max`
2. Separate updating of form data object from error message handling
   * Data inputted into input fields would instantly change data stored in the formData object using the _input_ event rather than waiting for the _blur_ event. This makes it work more like x-model.
   * Error messages will work like before: triggered initially on _blur_ event. If the _blur_ event triggers an error then it adds 'input' event listener for validation. This allows the error message to disappear as soon as the validation error is corrected.
3. Simplify the magic functions
   * Remove all validation functions and replace with optional pattern plugin for those that need this (see Breaking Changes 2 for more on this).
   * Basically thinking of simplifying it all so $validate.value() becomes $value() and etc.. Haven't put a lot of thought here but I'd like to streamline this.
4. Optional dual error messages for issue #11 with a simple required if left blank and more explicit message for failing pattern validation.

## Breaking Changes

1. Default to normal browser validation for input types email, tel, and url, rather than the added enhanced regex.
2. Only set x-validate on the form rather than individual fields; remove all x-validate field directives.
   * In general the field directives are all regex patterns which add unnecessary bulk to the plugin size. 
   * Replace this with an optional pattern plugin with magic functions. You can use these by binding to the pattern attribute: `:pattern="$pattern.email"`
   * x-validate.group is replaced (see Enhancement 1 above)
3. Remove the ability to use expressions for validation (i.e., `x-validate='$el.value === 'bunny'`)
   * Binding to the pattern attribute with normal AlpineJS will work for most if not all use cases. The plugin already checks validity including the pattern so _x-validate="expression"_ is superfluous.
   * For example: if you need to have a user add the same email twice you could use `:pattern="$validate.value('email')"`

Let me know what you think. Are there any enhancements you'd like to see? How terrible are the breaking changes for your use? Do you use the ability to use expressions for validation in a way that binding to the pattern attribute won't work?

**Comment by colinaut** [comment]
Tagging in some people who have added issues in the last year in order to elicit comments: @kevinmu17 @Rocketpilot @reatlat @64knl @QJan84 @g-rodigy @nomin-sasabuchi @basloki @phoozle @shopper360mst

Hope I'm not bugging y'all but I'd love feedback. Thanks!

**Comment by Rocketpilot** [comment]
"Default to normal browser validation for input types email" - I've been doing some reading on the brutally complex edge cases with email validation and I think punting that to the browser is an excellent idea. Let them take the heat for it.

**Comment by 64knl** [comment]
Really like the ideas, especially supporting multiple error messages!

Maybe I don't exactly get what you mean by defaulting to browser validation. I personally don't like defaulting to browser validation as this will be an inconsistent UI experience (and may display messages in a different language). Also I assumed browser validation would always kick in before any other validation.
I usually add browser validation as a fallback for all fields and then programmatically add `novalidate` to the form when using javascript validation.

**Reply by colinaut** [reply]
What I mean is defaulting to the browsers built in patterns for type="tel", type="email", and type="url". The x-validate would still set novalidate on the form so all error messages would be handled using x-validate.

Right now how x-validate does validity checks for those types of fields is:

1.  It runs `field.checkValidity()` which is the browser's normal check for those types and also works with any added pattern attribute regex. That `field.checkValidity()` does not trigger the browser's error popups it just returns a boolean.
2. It then automatically also runs x-validates enhanced "tel", "email", or "url" regex on top of that to ensure validity.
3. Then it runs any additional expressions for testing if that was added.
4. Then it triggers the error message if valid = false.

My plan for this plugin is to remove steps 2 and 3 as it can be handled by the pattern attribute. When I first designed this plugin I was unhappy with the browser's built in checks. The browser's built in is very broad. But since then I have realized how many crazy edge cases there are which is the reason that the browser's pattern for those is so broad.

I think it's better to by default just used the browser's validation and if any developer what's or needs enhanced regex for more specific validation they can use the pattern attribute. Not only is this more flexible but it is in keeping with the name of this plugin "Simple Validate" not to include additional stuff that may not be needed. The other benefit of using pattern attribute is if javascript fails then it still works since pattern attribute is basic browser tech.

**Reply by 64knl** [reply]
Thanks for the elaborate reply, there was a lot I didn't know about browser validation. Your solution sounds perfect. Looking forward to the new version!

**Comment by colinaut** [comment]
BTW. I am currently working on a web component called [Action Form](https://github.com/colinaut/action-form) which does a lot of what alpinejs-plugin-simple-validate does, without the need for AlpineJS. It's functionality isn't exactly the same — no magic functions or the like since it's not trying to tie into AlpineJS.

This doesn't mean I'm abandoning this plugin. I'll continue to support alpinejs-plugin-simple-validate. But it does put 2.0 of this plugin a bit on the back burner. This is actually a good thing however. Working on action-form has made me realize there may be better ways to handle some of the code I have in alpinejs-plugin-simple-validate.
